### ISSUE TITLE
Virtual - multiple rules generated for ...

### ISSUE THREAD

**Issue by nickbetteridge** [ocaml/dune#1895]
I've hit the following error and I'm not too sure on the best way forward for resolving it
```
dune build -j1 --root . @install
Multiple rules generated for _build/default/lib_dom_min/.ojs_subtlecrypto.objs/js_dom.pp.ml.all-deps:
- <internal location>
- src/virtual_rules.ml:39
Makefile:4: recipe for target 'build' failed
make: *** [build] Error 1
```
virtual_rules.ml must have been cleared up somewhere on exit.

These are the two dune files:
```
(library
  (name ojs_subtlecrypto_lib)
  (public_name ojs-subtlecrypto-lib)
  (virtual_modules dom)
  (modes byte)
  (libraries logs logs.fmt logs.browser fmt.tty win-error cstruct lwt gen_js_api)
  (preprocess (pps js_of_ocaml-ppx))
  (flags (:standard -w -7-26-27-32-33-53-55))
)
```
and
```
(library
  (name ojs_subtlecrypto)
  (public_name ojs-subtlecrypto)
  (implements ojs_subtlecrypto_lib)
  (private_modules js_dom)
  (libraries gen_js_api)
  (preprocess (pps js_of_ocaml-ppx))
  (flags (:standard -w -7-26-27-32-33-53-55))
)
```
The private_modules ```js_dom``` looks like it might be the trouble maker - how do I resolve the multiple rules error?

**Comment by rgrinberg** [comment]
Do you have a repo on github that I can use to reproduce? I'd like to create a test case out of this.

**Comment by rgrinberg** [comment]
If not, is the `js_dom` module defined in the `ojs_subtlecrypto_lib` library?

**Comment by nickbetteridge** [comment]
You're quite right - I was blind to an empty ```js_dom.ml``` in the lib - thanks for pointing out such an elementary mistake :) I've created ```https://github.com/nickbetteridge/ojs-subtlecrypto``` and will push everything into the repo - I still need to do a complete set of test cases and work the documentation. Thanks again

**Comment by rgrinberg** [comment]
Well, that's actually not a bug. You're allowed to have conflicting module names as long as they are private. And even if it wasn't allowed, whenever you see a bad error message like that, it's almost certainly unintended.

**Comment by rgrinberg** [comment]
Of course, in this case the best one could hope for is that you're shadowing the parent `js_dom` module. That is usually not intended, but nevertheless we don't want this to break stuff. E.g. if the virtual library adds a `Js_dom` module in a later version. You wouldn't want your implementation to just break.

**Comment by nickbetteridge** [comment]
It's not just a case of shadowing the Js_dom - you can tailor the Js_dom bindings to just the ones you want - saves downloading a large, complete set of dom bindings

**Comment by rgrinberg** [comment]
I tried reproducing this bug in https://github.com/ocaml/dune/pull/2710 and failed. I believe it is now fixed.
